File: src/main/java/net/dunciboy/discord_bot/commands/Ban.java
/*
 * MIT License
 *
 * Copyright (c) 2017 Duncan Casteleyn
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package net.dunciboy.discord_bot.commands;

import net.dunciboy.discord_bot.GuildLogger;
import net.dunciboy.discord_bot.JDALibHelper;
import net.dunciboy.discord_bot.RunBots;
import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.MessageBuilder;
import net.dv8tion.jda.core.Permission;
import net.dv8tion.jda.core.entities.*;
import net.dv8tion.jda.core.events.message.MessageReceivedEvent;
import net.dv8tion.jda.core.exceptions.PermissionException;
import net.dv8tion.jda.core.requests.RestAction;

import java.awt.*;
import java.util.concurrent.TimeUnit;

/**
 * Created by Duncan on 24/02/2017.
 * <p>
 * This class creates a command to ban users with logging.
 */
public class Ban extends CommandModule {
    private static final String[] ALIASES = new String[]{"Ban"};
    private static final String ARGUMENTATION_SYNTAX = "[User mention] [Reason~]";
    private static final String DESCRIPTION = "Will ban the mentioned user, clear all message that where posted by the user in the last 24 hours and log it to the log channel.";

    /**
     * Constructor for abstract class.
     */
    public Ban() {
        super(ALIASES, ARGUMENTATION_SYNTAX, DESCRIPTION);
    }

    /**
     * Constructor for abstract class
     *
     * @param aliases             the description for this command
     * @param argumentationSyntax the syntax for the argumentation of the command, put null if none needed.
     * @param description         The description of the command
     */
    Ban(String[] aliases, String argumentationSyntax, String description) {
        super(aliases, argumentationSyntax, description);
    }

    /**
     * Do something with the event, command and arguments.
     *
     * @param event     A MessageReceivedEvent that came with the command
     * @param command   The command alias that was used to trigger this commandExec
     * @param arguments The arguments that where entered after the command alias
     */
    @Override
    public void commandExec(MessageReceivedEvent event, String command, String arguments) {
        event.getAuthor().openPrivateChannel().queue(
                privateChannel -> commandExec(event, arguments, privateChannel),
                throwable -> commandExec(event, arguments, (PrivateChannel) null)
        );
    }

    void commandExec(MessageReceivedEvent event, String arguments, PrivateChannel privateChannel) {
        if (!event.isFromType(ChannelType.TEXT)) {
            if (privateChannel != null) {
                event.getChannel().sendMessage("This command only works in a guild.").queue(message -> message.delete().queueAfter(1, TimeUnit.MINUTES));
            }
        } else if (!event.getMember().hasPermission(Permission.BAN_MEMBERS)) {
            if (privateChannel != null) {
                privateChannel.sendMessage(event.getAuthor().getAsMention() + " you don't have permission to ban!").queue(message -> message.delete().queueAfter(1, TimeUnit.MINUTES));
            }
        } else if (event.getMessage().getMentionedUsers().size() < 1) {
            if (privateChannel != null) {
                privateChannel.sendMessage("Illegal argumentation, you need to mention a user that is still in the server.").queue();
            }
        } else {
            String reason;
            try {
                reason = arguments.substring(arguments.split(" ")[0].length() + 1);
            } catch (IndexOutOfBoundsException e) {
                throw new IllegalArgumentException("No reason provided for this action.");
            }
            Member toBan = event.getGuild().getMember(event.getMessage().getMentionedUsers().get(0));
            if (!event.getMember().canInteract(toBan)) {
                throw new PermissionException("You can't interact with this member");
            }
            RestAction<Void> banRestAction = event.getGuild().getController().ban(toBan, 1, reason);
            StringBuilder description = new StringBuilder("Reason: " + reason);
            if (event.getGuild().getIdLong() == 175856762677624832L) {
                description.append("\n\n")
                        .append("If you'd like to appeal the ban, please use this form: https://goo.gl/forms/SpWg49gaQlMt4lSG3");
                //todo make this configurable per guild.
            }
            MessageEmbed userKickNotification = new EmbedBuilder()
                    .setColor(Color.red)
                    .setAuthor(JDALibHelper.INSTANCE.getEffectiveNameAndUsername(event.getMember()), null, event.getAuthor().getEffectiveAvatarUrl())
                    .setTitle(event.getGuild().getName() + ": You have been banned by " + JDALibHelper.INSTANCE.getEffectiveNameAndUsername(event.getMember()), null)
                    .setDescription(description.toString())
                    .build();

            toBan.getUser().openPrivateChannel().queue(
                    privateChannelUserToMute -> privateChannelUserToMute.sendMessage(userKickNotification).queue(
                            message -> onSuccessfulInformUser(event, reason, privateChannel, toBan, message, banRestAction),
                            throwable -> onFailToInformUser(event, reason, privateChannel, toBan, throwable, banRestAction)
                    ),
                    throwable -> onFailToInformUser(event, reason, privateChannel, toBan, throwable, banRestAction)
            );
        }
    }

    private void logBan(MessageReceivedEvent event, String reason, Member toBan) {
        RunBots runBots = RunBots.Companion.getRunBot(event.getJDA());
        if (runBots != null) {
            EmbedBuilder logEmbed = new EmbedBuilder()
                    .setColor(Color.RED)
                    .setTitle("User was banned | Case: " + GuildLogger.Companion.getCaseNumberSerializable(event.getGuild().getIdLong()))
                    .addField("User", JDALibHelper.INSTANCE.getEffectiveNameAndUsername(toBan), true)
                    .addField("Moderator", JDALibHelper.INSTANCE.getEffectiveNameAndUsername(event.getMember()), true)
                    .addField("Reason", reason, false);

            runBots.getLogToChannel().log(logEmbed, toBan.getUser(), event.getGuild(), null);
        }
    }

    private void onSuccessfulInformUser(MessageReceivedEvent event, String reason, PrivateChannel privateChannel, Member toBan, Message userBanWarning, RestAction<Void> banRestAction) {
        banRestAction.queue(aVoid -> {
            logBan(event, reason, toBan);
            if (privateChannel == null) {
                return;
            }

            Message creatorMessage = new MessageBuilder()
                    .append("Banned ").append(toBan.toString()).append(".\n\nThe following message was sent to the user:")
                    .setEmbed(userBanWarning.getEmbeds().get(0))
                    .build();
            privateChannel.sendMessage(creatorMessage).queue();
        }, throwable -> {
            userBanWarning.delete().queue();
            if (privateChannel == null) {
                return;
            }

            Message creatorMessage = new MessageBuilder()
                    .append("Ban failed on ").append(toBan.toString())
                    .append(throwable.getClass().getSimpleName()).append(": ").append(throwable.getMessage())
                    .append(".\n\nThe following message was sent to the user but was automatically deleted:")
                    .setEmbed(userBanWarning.getEmbeds().get(0))
                    .build();
            privateChannel.sendMessage(creatorMessage).queue();
        });
    }

    private void onFailToInformUser(MessageReceivedEvent event, String reason, PrivateChannel privateChannel, Member toBan, Throwable throwable, RestAction<Void> banRestAction) {
        banRestAction.queue(aVoid -> {
            logBan(event, reason, toBan);
            if (privateChannel == null) {
                return;
            }

            Message creatorMessage = new MessageBuilder()
                    .append("Banned ").append(toBan.toString())
                    .append(".\n\nWas unable to send a DM to the user please inform the user manually, if possible.\n")
                    .append(throwable.getClass().getSimpleName()).append(": ").append(throwable.getMessage())
                    .build();
            privateChannel.sendMessage(creatorMessage).queue();
        }, banThrowable -> {
            if (privateChannel == null) {
                return;
            }

            Message creatorMessage = new MessageBuilder()
                    .append("Ban failed on ").append(toBan.toString())
                    .append("\n\nWas unable to ban the user\n")
                    .append(banThrowable.getClass().getSimpleName()).append(": ").append(banThrowable.getMessage())
                    .append(".\n\nWas unable to send a DM to the user.\n")
                    .append(throwable.getClass().getSimpleName()).append(": ").append(throwable.getMessage())
                    .build();
            privateChannel.sendMessage(creatorMessage).queue();
        });
    }
}

File: src/main/kotlin/be/duncanc/discordmodbot/bot/commands/Ban.kt
/*
 * Copyright 2018 Duncan Casteleyn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package be.duncanc.discordmodbot.bot.commands

import be.duncanc.discordmodbot.bot.services.GuildLogger
import be.duncanc.discordmodbot.bot.utils.extractReason
import be.duncanc.discordmodbot.bot.utils.findMemberAndCheckCanInteract
import be.duncanc.discordmodbot.bot.utils.nicknameAndUsername
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.channel.ChannelType
import net.dv8tion.jda.api.entities.channel.concrete.PrivateChannel
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.requests.RestAction
import net.dv8tion.jda.api.utils.messages.MessageCreateBuilder
import net.dv8tion.jda.api.utils.messages.MessageCreateData
import org.springframework.stereotype.Component
import java.awt.Color
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Created by Duncan on 24/02/2017.
 *
 *
 * This class creates a command to ban users with logging.
 */
@Component
class Ban : CommandModule(
    arrayOf("Ban"),
    "[User mention] [Reason~]",
    "Will ban the mentioned user, clear all message that where posted by the user in the last 24 hours and log it to the log channel.",
    true,
    true
) {

    override fun commandExec(event: MessageReceivedEvent, command: String, arguments: String?) {
        event.author.openPrivateChannel().queue(
            { privateChannel -> commandExec(event, arguments, privateChannel) }
        ) { commandExec(event, arguments, null as PrivateChannel?) }
    }

    fun commandExec(event: MessageReceivedEvent, arguments: String?, privateChannel: PrivateChannel?) {
        if (!event.isFromType(ChannelType.TEXT)) {
            if (privateChannel != null) {
                event.channel.sendMessage("This command only works in a guild.")
                    .queue { message -> message.delete().queueAfter(1, TimeUnit.MINUTES) }
            }
        } else if (event.member?.hasPermission(Permission.BAN_MEMBERS) != true) {
            privateChannel?.sendMessage(event.author.asMention + " you don't have permission to ban!")
                ?.queue { message -> message.delete().queueAfter(1, TimeUnit.MINUTES) }
        } else if (event.message.mentions.users.size < 1) {
            privateChannel?.sendMessage("Illegal argumentation, you need to mention a user that is still in the server.")
                ?.queue()
        } else {
            val reason: String = extractReason(arguments)

            val toBan = findMemberAndCheckCanInteract(event)
            val banRestAction = event.guild.ban(toBan, 7, TimeUnit.DAYS)
            val description = StringBuilder("Reason: $reason")
            if (event.guild.idLong == 175856762677624832L) {
                description.append("\n\n")
                    .append("If you'd like to appeal the ban, please use this form: https://goo.gl/forms/SpWg49gaQlMt4lSG3")
                //todo make this configurable per guild.
            } else if (event.guild.idLong == 176028172729450497L) {
                description.append("\n\n")
                    .append("If you'd like to appeal the ban, please use this form: https://forms.gle/ffbDj12KcSyTT7mUA")
                //todo make this configurable per guild.
            }
            val userBanNotification = EmbedBuilder()
                .setColor(Color.red)
                .setAuthor(event.member?.nicknameAndUsername, null, event.author.effectiveAvatarUrl)
                .setTitle(event.guild.name + ": You have been banned by " + event.member!!.nicknameAndUsername, null)
                .setDescription(description.toString())
                .build()

            toBan.user.openPrivateChannel().queue(
                { privateChannelUserToMute ->
                    privateChannelUserToMute.sendMessage(MessageCreateData.fromEmbeds(userBanNotification)).queue(
                        { message ->
                            onSuccessfulInformUser(
                                event,
                                reason,
                                privateChannel,
                                toBan,
                                message,
                                banRestAction
                            )
                        }
                    ) { throwable ->
                        onFailToInformUser(
                            event,
                            reason,
                            privateChannel,
                            toBan,
                            throwable,
                            banRestAction
                        )
                    }
                }
            ) { throwable -> onFailToInformUser(event, reason, privateChannel, toBan, throwable, banRestAction) }
        }
    }

    private fun logBan(event: MessageReceivedEvent, reason: String, toBan: Member) {
        val guildLogger = event.jda.registeredListeners.firstOrNull { it is GuildLogger } as GuildLogger?
        if (guildLogger != null) {
            val logEmbed = EmbedBuilder()
                .setColor(Color.RED)
                .setTitle("User banned")
                .addField("UUID", UUID.randomUUID().toString(), false)
                .addField("User", toBan.nicknameAndUsername, true)
                .addField("Moderator", event.member!!.nicknameAndUsername, true)
                .addField("Reason", reason, false)


            guildLogger.log(logEmbed, toBan.user, event.guild, null, GuildLogger.LogTypeAction.MODERATOR)
        }
    }

    private fun onSuccessfulInformUser(
        event: MessageReceivedEvent,
        reason: String,
        privateChannel: PrivateChannel?,
        toBan: Member,
        userBanWarning: Message,
        banRestAction: RestAction<Void>
    ) {
        banRestAction.queue({
            logBan(event, reason, toBan)
            if (privateChannel == null) {
                return@queue
            }

            val creatorMessage = MessageCreateBuilder()
                .addContent("Banned ").addContent(toBan.toString())
                .addContent(".\n\nThe following message was sent to the user:")
                .setEmbeds(userBanWarning.embeds)
                .build()
            privateChannel.sendMessage(creatorMessage).queue()
        }) { throwable ->
            userBanWarning.delete().queue()
            if (privateChannel == null) {
                return@queue
            }

            val creatorMessage = MessageCreateBuilder()
                .addContent("Ban failed on ").addContent(toBan.toString())
                .addContent(throwable.javaClass.simpleName).addContent(": ").addContent(throwable.message.toString())
                .addContent(".\n\nThe following message was sent to the user but was automatically deleted:")
                .setEmbeds(userBanWarning.embeds)
                .build()
            privateChannel.sendMessage(creatorMessage).queue()
        }
    }

    private fun onFailToInformUser(
        event: MessageReceivedEvent,
        reason: String,
        privateChannel: PrivateChannel?,
        toBan: Member,
        throwable: Throwable,
        banRestAction: RestAction<Void>
    ) {
        banRestAction.queue({
            logBan(event, reason, toBan)
            if (privateChannel == null) {
                return@queue
            }

            val creatorMessage = MessageCreateBuilder()
                .addContent("Banned ").addContent(toBan.toString())
                .addContent(".\n\nWas unable to send a DM to the user please inform the user manually, if possible.\n")
                .addContent(throwable.javaClass.simpleName).addContent(": ").addContent(throwable.message ?: "Unknown")
                .build()
            privateChannel.sendMessage(creatorMessage).queue()
        }) { banThrowable ->
            if (privateChannel == null) {
                return@queue
            }

            val creatorMessage = MessageCreateBuilder()
                .addContent("Ban failed on ").addContent(toBan.toString())
                .addContent("\n\nWas unable to ban the user\n")
                .addContent(banThrowable.javaClass.simpleName).addContent(": ")
                .addContent(banThrowable.message ?: "Unknown")
                .addContent(".\n\nWas unable to send a DM to the user.\n")
                .addContent(throwable.javaClass.simpleName).addContent(": ").addContent(throwable.message ?: "Unknown")
                .build()
            privateChannel.sendMessage(creatorMessage).queue()
        }
    }
}
